webhook/workspace/handler/validate.go

```diff
@@ -0,0 +1,105 @@
+//
+// Copyright (c) 2019-2021 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+package handler
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/devfile/devworkspace-operator/pkg/library/flatten"
+	registry "github.com/devfile/devworkspace-operator/pkg/library/flatten/internal_registry"
+
+	dwv2 "github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	devfilevalidation "github.com/devfile/api/v2/pkg/validation"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+func (h *WebhookHandler) ValidateDevfile(ctx context.Context, req admission.Request) admission.Response {
+
+	wksp := &dwv2.DevWorkspace{}
+	err := h.Decoder.Decode(req, wksp)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	workspace := &wksp.Spec.Template
+
+	// flatten the devworkspace if its not already flattened
+	if !flatten.DevWorkspaceIsFlattened(&wksp.Spec.Template) {
+		flattenHelpers := flatten.ResolverTools{
+			WorkspaceNamespace: wksp.Namespace,
+			Context:            ctx,
+			K8sClient:          h.Client,
+			InternalRegistry:   &registry.InternalRegistryImpl{},
+			HttpClient:         http.DefaultClient,
+		}
+		workspace, _, err = flatten.ResolveDevWorkspace(&wksp.Spec.Template, flattenHelpers)
+		if err != nil {
+			return admission.Errored(http.StatusInternalServerError, err)
+		}
+	}
+
+	commands := workspace.Commands
+	components := workspace.Components
+	events := workspace.Events
+	projects := workspace.Projects
+	starterProjects := workspace.StarterProjects
+
+	var devfileErrors []string
+
+	// validate commands
+	if commands != nil && components != nil {
+		cmdErrors := devfilevalidation.ValidateCommands(commands, components)
```

Comment:
well, we must think carefully about what and when we should validate... The corner case I have in mind. We verify and create DevWorkspace which is valid because it used volume from the plugin...

Then DWT is updated and volume is dropped.
DevWorkspace can't be started/stopped/updated because of DWT.

So, the state we can proceed with:
- drop all validation from webhooks except ones that can't use stuff from DWT, like projects;
- do whole validation on try to start a workspace;

Think more carefully about what we can validate in webhooks... Probably we should not do flattening here at all.

Reply:
I've switched to only validating projects, starterProjects, and events since none of those validations rely on DWT's and don't need flattening. I also find out that if you change `_store_tls_cert` to grab the `devworkspace-webhookserver-tls` secret on openshift then you don't need the additional scripts for generating the serving certs locally